contracts/multicall/GuardedMulticaller2.sol

```diff
@@ -0,0 +1,240 @@
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                _multicallSigner,
+                _hashTypedData(_reference, _calls, _deadline),
+                _signature
+            )
+        ) {
+            revert UnauthorizedSignature(_signature);
+        }
+
+        replayProtection[_reference] = true;
+
+        // Multicall
+        for (uint256 i = 0; i < _calls.length; i++) {
+            bytes4 functionSelector = bytes4(keccak256(bytes(_calls[i].functionSignature)));
+            bytes memory callData = abi.encodePacked(functionSelector, _calls[i].data);
+            // solhint-disable avoid-low-level-calls
+            // slither-disable-next-line calls-loop
+            (bool success, bytes memory returnData) = _calls[i].target.call(callData);
+            if (!success) {
+                // Look for revert reason and bubble it up if present
+                if (returnData.length == 0) {
+                    revert FailedCall(_calls[i]);
+                }
+                // solhint-disable-next-line no-inline-assembly
+                assembly {
+                    // The easiest way to bubble the revert reason is using memory via assembly
+                    revert(add(returnData, 32), mload(returnData))
+                }
+            }
+        }
+
+        emit Multicalled(_multicallSigner, _reference, _calls, _deadline);
+    }
+
+    // slither-disable-end low-level-calls,cyclomatic-complexity
+
+    /**
+     * @notice Grants MULTICALL_SIGNER_ROLE to a user. Only DEFAULT_ADMIN_ROLE can call this function.
+     *
+     * @param _user User to grant MULTICALL_SIGNER_ROLE to
+     */
+    function grantMulticallSignerRole(address _user) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        grantRole(MULTICALL_SIGNER_ROLE, _user);
+    }
+
+    /**
+     * @notice Revokes MULTICALL_SIGNER_ROLE for a user. Only DEFAULT_ADMIN_ROLE can call this function.
+     *
+     * @param _user User to grant MULTICALL_SIGNER_ROLE to
+     */
+    function revokeMulticallSignerRole(address _user) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        revokeRole(MULTICALL_SIGNER_ROLE, _user);
+    }
+
+    /**
+     * @notice Gets whether the reference has been executed before.
+     *
+     * @param _reference Reference to check
+     */
+    function hasBeenExecuted(bytes32 _reference) external view returns (bool) {
+        return replayProtection[_reference];
+    }
+
+    /**
+     *
+     * @dev Returns hash of array of calls
+     *
+     * @param _calls Array of calls
+     */
+    function _hashCallArray(Call[] calldata _calls) internal pure returns (bytes32) {
+        bytes32[] memory hashedCallArr = new bytes32[](_calls.length);
+        for (uint256 i = 0; i < _calls.length; i++) {
+            hashedCallArr[i] = keccak256(
+                abi.encode(CALL_TYPEHASH, _calls[i].target, _calls[i].functionSignature, _calls[i].data)
+            );
+        }
+        return keccak256(abi.encodePacked(hashedCallArr));
```

Comment:
Similar to comment above. Avoid using`encodePacked` with dynamic types.

Reply:
This makes sense. Using `encodePacked` in `keccak256` would cause a hash collision. Thanks!